scripts/run_integration_tests_aarch64.sh

```diff
@@ -0,0 +1,218 @@
+BIONIC_OS_QCOW2_UNCOMPRESSED_IMAGE="$WORKLOADS_DIR/$BIONIC_OS_QCOW2_IMAGE_UNCOMPRESSED_NAME"
+if [ ! -f "$BIONIC_OS_QCOW2_UNCOMPRESSED_IMAGE" ]; then
+    pushd $WORKLOADS_DIR
+    time qemu-img convert -p -f raw -O qcow2 $BIONIC_OS_RAW_IMAGE_NAME $BIONIC_OS_QCOW2_UNCOMPRESSED_IMAGE || exit 1
+    popd
+fi
+
+FOCAL_OS_IMAGE_DOWNLOAD_NAME="focal-server-cloudimg-arm64.img" \
+FOCAL_OS_IMAGE_DOWNLOAD_URL="https://cloudhypervisorstorage.blob.core.windows.net/images/$FOCAL_OS_IMAGE_DOWNLOAD_NAME"
+FOCAL_OS_DOWNLOAD_IMAGE="$WORKLOADS_DIR/$FOCAL_OS_IMAGE_DOWNLOAD_NAME"
+if [ ! -f "$FOCAL_OS_DOWNLOAD_IMAGE" ]; then
+    pushd $WORKLOADS_DIR
+    time wget --quiet $FOCAL_OS_IMAGE_DOWNLOAD_URL || exit 1
+    popd
+fi
+
+FOCAL_OS_RAW_IMAGE_NAME="focal-server-cloudimg-arm64.raw" \
+FOCAL_OS_RAW_IMAGE="$WORKLOADS_DIR/$FOCAL_OS_RAW_IMAGE_NAME"
+if [ ! -f "$FOCAL_OS_RAW_IMAGE" ]; then
+    pushd $WORKLOADS_DIR
+    time qemu-img convert -p -f qcow2 -O raw $FOCAL_OS_IMAGE_DOWNLOAD_NAME $FOCAL_OS_RAW_IMAGE_NAME || exit 1
+    popd
+fi
+
+# Convert the raw image to qcow2 image to remove compressed blocks from the disk. Therefore letting the
+# qcow2 format image can be directly used in the integration test.
+FOCAL_OS_QCOW2_IMAGE_UNCOMPRESSED_NAME="focal-server-cloudimg-arm64.qcow2" \
+FOCAL_OS_QCOW2_UNCOMPRESSED_IMAGE="$WORKLOADS_DIR/$FOCAL_OS_QCOW2_IMAGE_UNCOMPRESSED_NAME"
+if [ ! -f "$FOCAL_OS_QCOW2_UNCOMPRESSED_IMAGE" ]; then
+    pushd $WORKLOADS_DIR
+    time qemu-img convert -p -f raw -O qcow2 $FOCAL_OS_RAW_IMAGE_NAME $FOCAL_OS_QCOW2_UNCOMPRESSED_IMAGE || exit 1
+    popd
+fi
+
+ALPINE_MINIROOTFS_URL="http://dl-cdn.alpinelinux.org/alpine/v3.11/releases/aarch64/alpine-minirootfs-3.11.3-aarch64.tar.gz"
+ALPINE_MINIROOTFS_TARBALL="$WORKLOADS_DIR/alpine-minirootfs-aarch64.tar.gz"
+if [ ! -f "$ALPINE_MINIROOTFS_TARBALL" ]; then
+    pushd $WORKLOADS_DIR
+    time wget --quiet $ALPINE_MINIROOTFS_URL -O $ALPINE_MINIROOTFS_TARBALL || exit 1
+    popd
+fi
+
+ALPINE_INITRAMFS_IMAGE="$WORKLOADS_DIR/alpine_initramfs.img"
+if [ ! -f "$ALPINE_INITRAMFS_IMAGE" ]; then
+    pushd $WORKLOADS_DIR
+    mkdir alpine-minirootfs
+    tar xf "$ALPINE_MINIROOTFS_TARBALL" -C alpine-minirootfs
+    cat > alpine-minirootfs/init <<-EOF
+		#! /bin/sh
+		mount -t devtmpfs dev /dev
+		echo \$TEST_STRING > /dev/console
+		poweroff -f
+	EOF
+    chmod +x alpine-minirootfs/init
+    cd alpine-minirootfs
+    find . -print0 |
+        cpio --null --create --verbose --owner root:root --format=newc > "$ALPINE_INITRAMFS_IMAGE"
+    popd
+fi
+
+pushd $WORKLOADS_DIR
+sha1sum sha1sums-aarch64 --check
+if [ $? -ne 0 ]; then
+    echo "sha1sum validation of images failed, remove invalid images to fix the issue."
+    exit 1
+fi
+popd
+
+# Build custom kernel based on virtio-pmem and virtio-fs upstream patches
+PE_IMAGE="$WORKLOADS_DIR/Image"
+
+LINUX_CUSTOM_DIR="$WORKLOADS_DIR/linux-custom"
+
+if [ ! -f "$PE_IMAGE" ]; then
```

Comment:
Since these tests are run over and over again on the same machine you'll have problems if you update the kernel configuration or tree since as the old image file will be there. You should instead check if the directory is already present, if so update the branch, copy the config file and then trigger make. If nothing has changed then the build will complete quickly.

Reply:
Hi @rbradford ! I do understand that concern, but I think the current `run_integration_tests.sh` (even the x86_64 one) deletes the whole `linux-custom` directory after the kernel has been built and copied (see [here](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/b69f6d4f6cac6534a5f4575d768fa6fa5c401a0e/scripts/run_integration_tests.sh#L122)). I just followed this strategy for arm64. Therefore I am now a little bit confused. Do you prefer to change all of them? Thanks!

Reply:
The current integration tests are also kind of wrong in this regard. But it doesn't really matter as when as part of the CI we use a completely new machine so they have a new empty `workloads` directory. If anyone is running the integration tests manually and playing around with the kernel they are on their own.

But on the arm64 builder the workloads directory is preserved including any files in it (see https://github.com/cloud-hypervisor/cloud-hypervisor/pull/1395/files#diff-a2453ffa94fcffc7939322f32e4647afR319) so the only way to update the kernel will be to SSH in and manually delete `~/wokloads/Image` which is not sustainable.

I don't expect you to fix the x86_64 version of the script in your PR but you do need to do the right thing on the aarch64 one.

Reply:
Thanks @rbradford. Yes now I understand. There are two options for the aarch64 fix then:

1. Delete the existing `Image` everytime before we clone the linux source code.

2. Don't use the existing method for the linux source code directory, i.e. delete it after the kernel is built. Instead, we checking if there is any `linux-custom` directory first and then apply related right actions for either the directory is there or not.

Option 1 is easier but seems a temporary fix, and option 2 is more reasonable but I am not sure if the x86_64 will apply the same fix. Personally I would prefer keeping the arm64 and x86_64 in a same strategy (would be easier to maintain)...Any ideas?

Reply:
I would prefer option 2 and we can roll it out for the x86_64 script too. I think the best solution is to test if the mtime on the config file is newer than the kernel and if so force a rebuild of the kernel. There is no need to keep the kernel directory around.

You can do this with 
```bash
if [ "$PE_IMAGE" -ot "$SRCDIR/resources/linux-config-aarch64" ]; then
# Download kernel tree
# Copy config file
# Build kernel
# Copy into ~/workloads
fi
```

If we ever update the kernel without changing the kernel config (relatively unlikely) we can still force it with a whitespace change to the kernel config file.

Reply:
(You might also have to test if the file exist and or that it if it doesn't)

Reply:
@rbradford Done. Please check https://github.com/cloud-hypervisor/cloud-hypervisor/pull/1395/files#diff-9dd6649ac9e2a3df0cbc193cb3ec941fR119